_posts/2023-04-25-using-custom-principals-with-elytron.adoc

```diff
@@ -0,0 +1,283 @@
+        "process-state" => "reload-required"
+    }
+}
+
+[standalone@localhost:9990 /] reload
+----
+
+Now, whenever a user is authenticated via the `ApplicationDomain`, our
+`CustomPrincipal` will be associated with that user.
+
+== Accessing the custom principal
+
+There are many ways you can retrieve a principal from within an
+application. A few methods are listed below, with links to example
+applications you can try yourself.
+
+=== Jakarta Security
+
+With Jakarta Security, you can access the current user via the
+`SecurityContext` object. Inject the `SecurityContext` into your Jakarta
+Servlet, and use the standard methods `getCallerPrincipal()` or
+`getPrincipalsByType()` to retrieve the custom principal:
+
+.Accessing a custom principal with Jakarta Security
+[source,java]
+----
+package com.company.servlet;
+
+// [...]
+
+@WebServlet
+public class MyServlet extends HttpServlet {
+
+    @Inject
+    private SecurityContext securityContext;
+
+    private CustomPrincipal getCustomPrincipal() {
+        Principal custPrincipal = securityContext.getCallerPrincipal();
+        return (CustomPrincipal) custPrincipal;
+    }
+
+    private CustomPrincipal getCustomPrincipalByType() {
+        Set<CustomPrincipal> principals = securityContext.getPrincipalsByType(CustomPrincipal.class);
+        return principals.iterator().next();
+    }
+
+    // [...]
+}
+----
+
+To use this functionality, simply enable a default _Jakarta
+Authorization (JACC)_ policy in Elytron:
+
+[source,shell]
+----
+[standalone@localhost:9990 /] /subsystem=elytron/policy=jacc:add(jacc-policy={})
+----
+
+The https://github.com/wildfly-security-incubator/elytron-examples/tree/main/custom-principal-ee[*custom-principal-ee*]
+example is a full _Jakarta Security_ implementation. It demonstrates how
+both of these methods return the same class, making the custom principal
+available to the Servlet.
+
+=== Jakarta Enterprise Beans (EJBs)
+
+The custom principal can be retrieved by any class implementing
+`EJBContext`. For example, a stateless EJB can inject `SessionContext`, and
+call `getCallerPrincipal()` to retrieve the custom principal:
+
+.Accessing a custom principal from an EJB
+[source,java]
+----
+package com.company.beans;
+
+// [...]
+
+@Stateless
+@Remote(MyBeanInterface.class)
+public class MyBean implements MyBeanInterface {
+
+    @Resource
+    private SessionContext ejbContext;
+
+    @Override
+    public CustomPrincipal getCustomPrincipal() {
+        Principal custPrincipal = ejbContext.getCallerPrincipal();
+        return (CustomPrincipal) custPrincipal;
+    }
+
+    // [...]
+}
+----
+
+The https://github.com/wildfly-security-incubator/elytron-examples/tree/main/custom-principal-ejb[*custom-principal-ejb*]
+example demonstrates a pair of EJBs and a remote client using methods
+from a custom principal.
+
+=== Using SecurityContext with Elytron
+
+When securing an application using an Elytron HTTP authentication
```

Comment:
@cam-rod  Just a minor nitpick. Since we are referring to the `Jakarta Security` and mention `SecurityContext` here, maybe it would be better to put it after the Jakarta Security chapter? So in the below order:

1.Jakarta Security
2. Using SecurityContext with Elytron
3. Jakarta Enterprise Beans (EJBs)
 
if that makes sense to you as well

Reply:
Yeah that makes sense, since the EJB demo also uses Elytron's authentication process. Swapping them around